### ISSUE TITLE
I think this program has a memory leak problem

### ISSUE THREAD

**Issue by antibodywzy** [tejasmorkar/sketch-to-color#2]
Hi Tejas,

Your article and program has a great all around explanation. Thanks for a great write-up.

I am a beginner in deep learning, and I also like anime very much so I can’t wait to download and try to run it on my computer after seeing your article. However, When I try to run it on my computer, it often suffers from memory leaks. Generally, an error ‘Resource Exhausted' will be reported after running for about 10 epochs. I don’t think this is due to the performance of my computer. My computer has a GTX 1070 and 16 Gb memory. I even rent a cloud server which contains two RTX 2080ti and 64Gb memory, It still occurred. 

I don't know the reason of this. Can you check if there is a memory leak problem in this code?

Best.

Bigpaper

**Comment by tejasmorkar** [comment]
Hi @antibodywzy ,

I am glad that you found the explanation helpful from the article.

This code worked on my system for approximately 150 epochs. My laptop has GTX 1060 6B and a 16 GB RAM. 
So, it should've worked on your computer fine but there may be some problems which I will definitely look into. 

Meanwhile, could you please send me the exact code you copied and the error you get. It might help me understand it better. 
Also, send me the versions numbers of python, tensorflow, Cuda toolkit, CuDNN and other related packages that you are using.

**Comment by antibodywzy** [comment]
Hi Tejas,

I download all the code from github page. So I don't think it was caused by copying the wrong code.

And the error occurred was

"2020-08-19 19:42:59.234899: W tensorflow/core/framework/op_kernel.cc:1767] OP_REQUIRES failed at image_resizer_state.h:142 : Resource exhausted: OOM when allocating tensor with shape[1,286,286,3] and type float on /job:localhost/replica:0/task:0/device:CPU:0 by allocator cpu"

I think it's a memory leak problem, but I don't know the reason. My python version is 3.8.3. Cuda toolkit is 10.2, and cudnn's version is matched with cuda. All other packages used are the latest version from anaconda.

Best.
Antibodywzy.

**Comment by tejasmorkar** [comment]
Hi @antibodywzy ,

I went through all the steps for again running the code and it works fine on small buffer_size. 
I was able to recreate the error when the buffer_size was 14224. My training was interrupted while shuffling on the 12th epoch. I agree that this shouldn't be the case. I went through the code and I did not see any reason for memory leakage due to the code. I have opened an issue on the TensorFlow repository and I'll keep you posted with any findings that I come across.

**Comment by tejasmorkar** [comment]
Hey @antibodywzy 

According to the comment on the [issue that I opened on the TensorFlow repo](https://github.com/tensorflow/tensorflow/issues/42586), using tcmalloc can help with the memory leak many times. 

The comment was:
`This may be due to heap fragmentation. tcmalloc does a much better job of avoiding fragmentation. There are some instructions here for how to switch to tcmalloc: https://riptutorial.com/tensorflow/example/13427/use-the-tcmalloc-allocator`
